<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Week 3 – Visualizing tabular data with ggplot2</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="styles.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"><img style="width: 120px; float: left; margin-top: 0px; margin-right: 20px;" src="images/CRUK_CI_logo.png"/>Introduction to R&nbsp;&nbsp;&nbsp;</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Contents
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="week0.html">Week 0 - Getting ready</a>
    </li>
    <li>
      <a href="week1.html">Week 1 - Introduction to R and RStudio</a>
    </li>
    <li>
      <a href="week2.html">Week 2 - Working with data in R</a>
    </li>
    <li>
      <a href="week3.html">Week 3 - Data visualization with ggplot2</a>
    </li>
    <li>
      <a href="week4.html">Week 4 - Data manipulation using dplyr</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/bioinformatics-core-shared-training/r-intro">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Week 3 – Visualizing tabular data with ggplot2</h1>

</div>


<blockquote>
<h4 id="learning-objectives">Learning objectives</h4>
<ul>
<li>Understand ggplot2’s grammar of graphics</li>
<li>Apply the grammar to create a variety of plots including scatter plots, box plots and histograms</li>
<li>Learn about how R handles categorical data</li>
</ul>
</blockquote>
<div id="visualizing-data-with-ggplot2" class="section level1">
<h1>Visualizing data with ggplot2</h1>
<p>In this session we will focus our attention on the popular <strong>ggplot2</strong> package for visualizing tabular data and will be learning how to create various different kinds of plot.</p>
<div id="metabric-data-set" class="section level2">
<h2>METABRIC data set</h2>
<p>We’ll be using an extended version of the METABRIC data set (from last week) in which columns have been added for the mRNA expression values for selected genes, including estrogen receptor alpha (ESR1), progesterone receptor (PGR), GATA3 and FOXA1.</p>
<p>The METABRIC study characterized the genomic mutations and gene expression profiles for over 2000 primary breast tumours. In addition to the gene expression data generated using microarrays, genome-wide copy number profiles were obtained using SNP microarrays and targeted sequencing was performed using a panel of 40 driver-mutation genes to detect mutations (single nucleotide variants).</p>
<p><a href="https://pubmed.ncbi.nlm.nih.gov/22522925">Curtis <em>et al.</em>, Nature 486:346-52, 2012</a></p>
<p><a href="https://www.ncbi.nlm.nih.gov/pubmed/27161491">Pereira <em>et al.</em>, Nature Communications 7:11479, 2016</a></p>
<p>Both the clinical data and the gene expression values were downloaded from <a href="https://www.cbioportal.org/study/summary?id=brca_metabric">cBioPortal</a>.</p>
<p>These have been combined into a single table using techniques we will cover later in the course. We’ve removed observations for patient tumour samples for which expression data are not available, so our table has fewer rows.</p>
<p>In addition, we have replaced spaces in column names with underscores so we don’t have to constantly refer to these using backticks ( <strong><code>`</code></strong> ). Note, however, that even after renaming some of the column names, we have one column, ‘3-gene classifier’, that is not strictly a valid name in R as it starts with a number and contains a hyphen. Valid names for variables in R contain only letters, numbers and the dot (<strong><code>.</code></strong>) and underscore (<strong><code>_</code></strong>) characters and should not begin with a number. So when we refer to the ‘3-gene classifier’ column we’ll have to use backticks, e.g. <code>metabric$`3-gene_classifier`</code>.</p>
<pre class="r"><code>library(tidyverse)
metabric &lt;- read_csv(&quot;data/metabric_clinical_and_expression_data.csv&quot;)
metabric</code></pre>
<pre><code>## # A tibble: 1,904 x 30
##    Patient_ID Cohort Age_at_diagnosis Survival_time Survival_status
##    &lt;chr&gt;       &lt;dbl&gt;            &lt;dbl&gt;         &lt;dbl&gt; &lt;chr&gt;          
##  1 MB-0000         1             75.6         140.  LIVING         
##  2 MB-0002         1             43.2          84.6 LIVING         
##  3 MB-0005         1             48.9         164.  DECEASED       
##  4 MB-0006         1             47.7         165.  LIVING         
##  5 MB-0008         1             77.0          41.4 DECEASED       
##  6 MB-0010         1             78.8           7.8 DECEASED       
##  7 MB-0014         1             56.4         164.  LIVING         
##  8 MB-0022         1             89.1          99.5 DECEASED       
##  9 MB-0028         1             86.4          36.6 DECEASED       
## 10 MB-0035         1             84.2          36.3 DECEASED       
## # … with 1,894 more rows, and 25 more variables: Vital_status &lt;chr&gt;,
## #   Chemotherapy &lt;chr&gt;, Radiotherapy &lt;chr&gt;, Tumour_size &lt;dbl&gt;,
## #   Tumour_stage &lt;dbl&gt;, Neoplasm_histologic_grade &lt;dbl&gt;,
## #   Cancer_type &lt;chr&gt;, ER_status &lt;chr&gt;, PR_status &lt;chr&gt;,
## #   HER2_status &lt;chr&gt;, HER2_status_measured_by_SNP6 &lt;chr&gt;, PAM50 &lt;chr&gt;,
## #   `3-gene_classifier` &lt;chr&gt;, Nottingham_prognostic_index &lt;dbl&gt;,
## #   Cellularity &lt;chr&gt;, Integrative_cluster &lt;chr&gt;, Mutation_count &lt;dbl&gt;,
## #   ESR1 &lt;dbl&gt;, ERBB2 &lt;dbl&gt;, PGR &lt;dbl&gt;, TP53 &lt;dbl&gt;, PIK3CA &lt;dbl&gt;,
## #   GATA3 &lt;dbl&gt;, FOXA1 &lt;dbl&gt;, MLPH &lt;dbl&gt;</code></pre>
</div>
</div>
<div id="our-first-ggplot---a-scatter-plot" class="section level1">
<h1>Our first ggplot - a scatter plot</h1>
<p>First we’ll create a simple scatter plot much like the ones we’ve created in previous sessions using R’s <code>plot()</code> function. We’ll plot the expression of estrogen receptor alpha (ESR1) against that of the transcription factor, GATA3.</p>
<pre class="r"><code>ggplot(data = metabric) +
  geom_point(mapping = aes(x = GATA3, y = ESR1))</code></pre>
<p><img src="week3_files/figure-html/scatter_plot_1-1.png" width="672" /></p>
<p>We specified 3 things to create this plot:</p>
<ol style="list-style-type: decimal">
<li>The <strong>data</strong> – needs to be a data frame (or a tibble)</li>
<li>The type of plot – this is called a <strong>geom</strong> in ggplot2 terminology</li>
<li>The mapping of variables in the data to visual properties of objects in the plot - these are called <strong>aesthetics</strong> in ggplot2</li>
</ol>
<p>In this case, the type of plot is a <strong><code>geom_point</code></strong>, ggplot2’s function for a scatter plot, and the expression of the genes GATA3 and ESR1 (estrogen receptor alpha) are mapped to the <strong><code>x</code></strong> and <strong><code>y</code></strong> coordinates, i.e. the positions of points on the scatter plot.</p>
<p>Other aesthetics include the size, shape and colour of the points. It might seem a little surprising that the x and y coordinates are treated in the same way as we wouldn’t normally think about these as having aesthetic qualities but this is indeed the case with ggplot2.</p>
</div>
<div id="aesthetic-mappings" class="section level1">
<h1>Aesthetic mappings</h1>
<p>Look up the <code>geom_point</code> help page to see the list of aesthetics that are available for this geom.</p>
<pre class="r"><code>?geom_point</code></pre>
<p>If you scroll down to the aesthetics section in the help page for <code>geom_point</code>, you’ll see that the <strong><code>x</code></strong> and <strong><code>y</code></strong> aesthetics are in bold face which tells us that these are required, i.e. you have to supply them, otherwise you’ll get an error message from ggplot.</p>
<p>Other aesthetics such as size, shape, colour and transparency (<code>alpha</code>) are optional.</p>
<p>So what about those other aesthetics? Colours are usually fun – let’s try to use the <code>colour</code> aesthetic.</p>
<p>We need to map a variable to a colour just like we did for x and y. We’ll use the ER status as an example.</p>
<pre class="r"><code>ggplot(data = metabric) +
  geom_point(mapping = aes(x = GATA3, y = ESR1, colour = ER_status))</code></pre>
<p><img src="week3_files/figure-html/scatter_plot_2-1.png" width="672" /></p>
<p>What about colouring points based on a continuous variable?</p>
<pre class="r"><code>ggplot(data = metabric) +
  geom_point(mapping = aes(x = GATA3, y = ESR1, colour = Nottingham_prognostic_index))</code></pre>
<p><img src="week3_files/figure-html/scatter_plot_3-1.png" width="672" /></p>
<p>ggplot2 uses a colour scale for a continous variable but discrete colours for discrete or categorical values.</p>
<p>Size is not usually a good aesthetic to map to a variable. This is particularly the case where we have such a large number of observations.</p>
<pre class="r"><code>ggplot(data = metabric) +
  geom_point(mapping = aes(x = GATA3, y = ESR1, size = ER_status))</code></pre>
<pre><code>## Warning: Using size for a discrete variable is not advised.</code></pre>
<p><img src="week3_files/figure-html/scatter_plot_4-1.png" width="672" /></p>
<p>Some aesthetics, such as size, are designed to be used with continuous variables. If you map these to a discrete variable as we have done here, then ggplot warns you that this isn’t such a good idea.</p>
<p>It would however be useful to be able to decrease the size of the points given that there are around 2000 in this data set. We can do so by setting the size to a single value rather than mapping it to one of the variables in the data set - this has to be done outside the aesthetic mappings, i.e. outside the <code>aes()</code> bit.</p>
<pre class="r"><code>ggplot(data = metabric) +
  geom_point(mapping = aes(x = GATA3, y = ESR1, colour = Nottingham_prognostic_index), size = 0.5)</code></pre>
<p><img src="week3_files/figure-html/scatter_plot_5-1.png" width="672" /></p>
<p>Some aesthetics can only be used with categorical variables, e.g. shape.</p>
<pre class="r"><code>ggplot(data = metabric) +
  geom_point(mapping = aes(x = GATA3, y = ESR1, shape = Survival_time))</code></pre>
<pre><code>Error: A continuous variable can not be mapped to shape</code></pre>
<p>Here we set the shape based on the 3-gene classifier.</p>
<pre class="r"><code>ggplot(data = metabric) +
  geom_point(mapping = aes(x = GATA3, y = ESR1, shape = `3-gene_classifier`), size = 0.9)</code></pre>
<pre><code>## Warning: Removed 204 rows containing missing values (geom_point).</code></pre>
<p><img src="week3_files/figure-html/scatter_plot_6-1.png" width="672" /></p>
<p>Note that some of the patient samples have not been classified and ggplot has removed those points with missing values for the 3-gene classifier.</p>
<p>It is usually preferable to use colours to distinguish between different categories but sometimes colour and shape are used together where we want to show which group a data point belongs to in two different categorical variables.</p>
<pre class="r"><code>ggplot(data = metabric) +
  geom_point(mapping = aes(x = GATA3, y = ESR1, colour = PAM50, shape = `3-gene_classifier`), size = 0.9)</code></pre>
<pre><code>## Warning: Removed 204 rows containing missing values (geom_point).</code></pre>
<p><img src="week3_files/figure-html/scatter_plot_7-1.png" width="672" /></p>
<p>Transparency can be useful when we have a large number of points as we can more easily tell when points are overlaid, but like size, it is not usually mapped to a variable and sits outside the <code>aes()</code>.</p>
<pre class="r"><code>ggplot(data = metabric) +
  geom_point(mapping = aes(x = GATA3, y = ESR1, colour = `3-gene_classifier`), size = 0.75, alpha = 0.5)</code></pre>
<p><img src="week3_files/figure-html/scatter_plot_8-1.png" width="672" /></p>
<p>Colours are more commonly used than shapes, sizes or transparency.</p>
<p>Virtually every aspect of the plots we’ve created can be customized. We will learn how to change the colours used as well as other display parameters later in the course.</p>
</div>
<div id="ggplot2-grammar" class="section level1">
<h1>ggplot2 grammar</h1>
<p>The “gg” in ggplot2 stands for “grammar of graphics”. This grammar is a coherent system for describing and building graphs.</p>
<p>The basic template is given in the summary box below. This template can be used to create any of a wide range of graph types with ggplot2.</p>
<div class="rmdblock">
<p>
<strong>ggplot2 grammar of graphics</strong>
</p>
<p>
<strong><code>ggplot(data = &lt;DATA&gt;) + &lt;GEOM_FUNCTION&gt;(mapping = aes(&lt;MAPPINGS&gt;))</code></strong>
</p>
<p>
<code>ggplot()</code> creates the plot object.
</p>
<p>
Layers can be added by adding geoms such as <code>geom_point()</code> using the <code>+</code> operator that ggplot2 has overridden for this purpose.
</p>
</div>
</div>
<div id="another-plot-type---bar-chart" class="section level1">
<h1>Another plot type - bar chart</h1>
<p>The range of geoms available in ggplot2 can be obtained by navigating to the ggplot2 package in the Packages tab pane in RStudio (bottom right-hand corner) and scrolling down the list of functions sorted alphabetically to the <code>geom_...</code> functions.</p>
<p>The METABRIC study redefined how we think about breast cancer by identifying and characterizing several new subtypes, referred to as <em>integrative clusters</em>. Let’s create a bar chart of the number of patients whose cancers fall within each subtype in the METABRIC cohort.</p>
<p><code>geom_bar</code> is the geom we need and it requires a single aesthetic mapping of the categorical variable of interest to <code>x</code>.</p>
<pre class="r"><code>ggplot(data = metabric) +
  geom_bar(mapping = aes(x = Integrative_cluster))</code></pre>
<p><img src="week3_files/figure-html/bar_plot_1-1.png" width="672" /></p>
<p>The dark grey bars are a big ugly - what if we want each bar to be a different colour?</p>
<pre class="r"><code>ggplot(data = metabric) +
  geom_bar(mapping = aes(x = Integrative_cluster, colour = Integrative_cluster))</code></pre>
<p><img src="week3_files/figure-html/bar_plot_2-1.png" width="672" /></p>
<p>Colouring the edges wasn’t quite what we had in mind. Look at the help for <code>geom_bar</code> to see what other aesthetic we should have used.</p>
<pre class="r"><code>ggplot(data = metabric) +
  geom_bar(mapping = aes(x = Integrative_cluster, fill = Integrative_cluster))</code></pre>
<p><img src="week3_files/figure-html/bar_plot_3-1.png" width="672" /></p>
<p>What happens if we colour (fill) with something other than the integrative cluster?</p>
<pre class="r"><code>ggplot(data = metabric) +
  geom_bar(mapping = aes(x = Integrative_cluster, fill = ER_status))</code></pre>
<p><img src="week3_files/figure-html/bar_plot_4-1.png" width="672" /></p>
<p>We get a stacked bar plot.</p>
<p>Note the similarity in what we did here to what we did with the scatter plot - there is a common grammar.</p>
<p>Let’s try another stacked bar plot, this time with a categorical variable with more than two categories.</p>
<pre class="r"><code>ggplot(data = metabric) +
  geom_bar(mapping = aes(x = Integrative_cluster, fill = `3-gene_classifier`))</code></pre>
<p><img src="week3_files/figure-html/bar_plot_5-1.png" width="672" /></p>
<p>What if want all the bars to be the same colour but not dark grey, e.g. blue?</p>
<pre class="r"><code>ggplot(data = metabric) +
  geom_bar(mapping = aes(x = Integrative_cluster, fill = &quot;blue&quot;))</code></pre>
<p><img src="week3_files/figure-html/bar_plot_6-1.png" width="672" /></p>
<p>That doesn’t look right - why not?</p>
<p>You can set the aesthetics to a fixed value but this needs to be outside the mapping, just like we did before for size and transparency in the scatter plots.</p>
<pre class="r"><code>ggplot(data = metabric) +
  geom_bar(mapping = aes(x = Integrative_cluster), fill = &quot;blue&quot;)</code></pre>
<p><img src="week3_files/figure-html/bar_plot_7-1.png" width="672" /></p>
<p>Setting this inside the <code>aes()</code> mapping told ggplot2 to map the colour aesthetic to some variable in the data frame, one that doesn’t really exist but which is created on-the-fly with a value of “blue” for every observation.</p>
<div id="statistical-transformations-advanced" class="section level2">
<h2>Statistical transformations (ADVANCED)</h2>
<p>You may have noticed that ggplot2 didn’t just plot values from our data set but had to do some calculation first for the bar chart, i.e. it had to sum the number of observations in each category.</p>
<p>Each geom has a <strong>statistical transformation</strong>. In the case of the scatter plot, <code>geom_point</code> uses the “identity” transformation which means just use the values as they are, i.e. not really a transformation at all. The statistical transformation for <code>geom_bar</code> is “count”, which means it will count the number of observations for each category in the variable mapped to the x aesthetic.</p>
<p>You can see which statistical transformation is being used by a geom by looking at the <code>stat</code> argument in the help page for that geom.</p>
<p>There are some circumstances where you’d want to change the <code>stat</code>, for example if we already had count values in our table.</p>
<pre class="r"><code>counts &lt;- tibble(
  `3-gene_classifier` = c(&quot;ER-/HER2-&quot;, &quot;ER+/HER2- High Prolif&quot;, &quot;ER+/HER2- Low Prolif&quot;, &quot;HER2+&quot;),
  count = c(290, 603, 619, 188)
)
counts</code></pre>
<pre><code>## # A tibble: 4 x 2
##   `3-gene_classifier`   count
##   &lt;chr&gt;                 &lt;dbl&gt;
## 1 ER-/HER2-               290
## 2 ER+/HER2- High Prolif   603
## 3 ER+/HER2- Low Prolif    619
## 4 HER2+                   188</code></pre>
<pre class="r"><code>ggplot(data = counts) +
  geom_bar(mapping = aes(x = `3-gene_classifier`, y = count), stat = &quot;identity&quot;)</code></pre>
<p><img src="week3_files/figure-html/bar_plot_8-1.png" width="672" /></p>
</div>
</div>
<div id="multiple-layers" class="section level1">
<h1>Multiple layers</h1>
<p>Consider again the ggplot2 grammar:</p>
<pre><code>ggplot(data = &lt;DATA&gt;) +
  &lt;GEOM_FUNCTION&gt;(mapping = aes(&lt;MAPPINGS&gt;))</code></pre>
<p>Why do we have the ggplot part and geom part joined by a “+” symbol?</p>
<p>The <strong>‘+’</strong> operator has been overridden by ggplot2 to add layers (geoms) and other components (scales, themes, etc.) to the plot.</p>
<p>We can have multiple geoms in the same plot. Each is a different layer.</p>
<p>Let’s create a plot with two geoms.</p>
<pre class="r"><code>ggplot(data = metabric) +
  geom_point(mapping = aes(x = GATA3, y = ESR1)) +
  geom_smooth(mapping = aes(x = GATA3, y = ESR1))</code></pre>
<pre><code>## `geom_smooth()` using method = &#39;gam&#39; and formula &#39;y ~ s(x, bs = &quot;cs&quot;)&#39;</code></pre>
<p><img src="week3_files/figure-html/scatter_line_plot_1-1.png" width="672" /></p>
<p>What is <strong><code>geom_smooth()</code></strong> doing? Look up the help page.</p>
<p>Note that the shaded area surrounding blue line represents the standard error bounds on the fitted model.</p>
<p>It is important when breaking these ggplot statements across multiple lines to ensure that the <code>+</code> is at the end of the line so that R knows you haven’t finished constructing your plot.</p>
<p>There is some annoying duplication of code used to create this plot. We’ve repeated the exact same aesthetic mapping for both geoms. We can avoid this by putting the mappings in the <code>ggplot()</code> function instead.</p>
<pre class="r"><code>ggplot(data = metabric, mapping = aes(x = GATA3, y = ESR1)) +
  geom_point() +
  geom_smooth()</code></pre>
<pre><code>## `geom_smooth()` using method = &#39;gam&#39; and formula &#39;y ~ s(x, bs = &quot;cs&quot;)&#39;</code></pre>
<p><img src="week3_files/figure-html/scatter_line_plot_2-1.png" width="672" /></p>
<p>Let’s make the plot look a bit prettier by reducing the size of the points and making them transparent. We’re not mapping <code>size</code> or <code>alpha</code> to any variables, just setting them to constant values, and we only want these settings to apply to the points, so we set them inside <code>geom_point()</code>.</p>
<pre class="r"><code>ggplot(data = metabric, mapping = aes(x = GATA3, y = ESR1)) +
  geom_point(size = 0.5, alpha = 0.5) +
  geom_smooth()</code></pre>
<pre><code>## `geom_smooth()` using method = &#39;gam&#39; and formula &#39;y ~ s(x, bs = &quot;cs&quot;)&#39;</code></pre>
<p><img src="week3_files/figure-html/scatter_line_plot_3-1.png" width="672" /></p>
<p>Aesthetics set in the <code>ggplot</code> function are treated as global mappings for the plot and are inherited by all geoms added to the plot.</p>
<p>Let’s say we still want to colour the points by ER status.</p>
<pre class="r"><code>ggplot(data = metabric, mapping = aes(x = GATA3, y = ESR1, colour = ER_status)) +
  geom_point(size = 0.5, alpha = 0.5) +
  geom_smooth()</code></pre>
<pre><code>## `geom_smooth()` using method = &#39;gam&#39; and formula &#39;y ~ s(x, bs = &quot;cs&quot;)&#39;</code></pre>
<p><img src="week3_files/figure-html/scatter_line_plot_4-1.png" width="672" /></p>
<p>Well, that wasn’t expected, although it’s pretty neat. We only wanted to apply the colour aesthetic to the points.</p>
<p>We can add the colour aesthetic to the <code>geom_point()</code> function instead of as a global mapping in the <code>ggplot()</code> function.</p>
<pre class="r"><code>ggplot(data = metabric, mapping = aes(x = GATA3, y = ESR1)) +
  geom_point(mapping = aes(colour = ER_status), size = 0.5, alpha = 0.5) +
  geom_smooth()</code></pre>
<pre><code>## `geom_smooth()` using method = &#39;gam&#39; and formula &#39;y ~ s(x, bs = &quot;cs&quot;)&#39;</code></pre>
<p><img src="week3_files/figure-html/scatter_line_plot_5-1.png" width="672" /></p>
<p>Or suppose you’ve spent a bit of time getting your scatter plot just right and decide to add another layer but you’re a bit worried about interfering with the the code you so lovingly crafted, you can set the <code>inherit.aes</code> option to <code>FALSE</code> and set the aesthetic mappings explicitly for your new layer.</p>
<pre class="r"><code>ggplot(data = metabric, mapping = aes(x = GATA3, y = ESR1, colour = ER_status)) +
  geom_point(size = 0.5, alpha = 0.5) +
  geom_smooth(mapping = aes(x = GATA3, y = ESR1), inherit.aes = FALSE)</code></pre>
<pre><code>## `geom_smooth()` using method = &#39;gam&#39; and formula &#39;y ~ s(x, bs = &quot;cs&quot;)&#39;</code></pre>
<p><img src="week3_files/figure-html/scatter_line_plot_6-1.png" width="672" /></p>
</div>
<div id="box-plots" class="section level1">
<h1>Box plots</h1>
<p>Box plots are a particular favourite of Wednesday lunchtime seminars.</p>
<pre class="r"><code>ggplot(data = metabric, mapping = aes(x = ER_status, y = GATA3)) +
  geom_boxplot()</code></pre>
<p><img src="week3_files/figure-html/box_plot_1-1.png" width="672" /></p>
<p>See <code>geom_boxplot</code> help to explain how the box and whiskers are constructed and how it decides which points are outliers and should be displayed as points.</p>
<p>How about adding another layer to display all the points?</p>
<pre class="r"><code>ggplot(data = metabric, mapping = aes(x = ER_status, y = GATA3)) +
  geom_boxplot() +
  geom_point()</code></pre>
<p><img src="week3_files/figure-html/box_plot_2-1.png" width="672" /></p>
<p>Ideally, we’d like these points to be spread out a bit. The <code>geom_point</code> help points to <code>geom_jitter</code> as more suitable when one of the variables is categorical.</p>
<pre class="r"><code>ggplot(data = metabric, mapping = aes(x = ER_status, y = GATA3)) +
  geom_boxplot() +
  geom_jitter()</code></pre>
<p><img src="week3_files/figure-html/box_plot_3-1.png" width="672" /></p>
<p>Well, that’s a bit of a mess. We need to reduce the spread or jitter and make the points smaller and transparent.</p>
<pre class="r"><code>ggplot(data = metabric, mapping = aes(x = ER_status, y = GATA3)) +
  geom_boxplot() +
  geom_jitter(width = 0.3, size = 0.5, alpha = 0.25)</code></pre>
<p><img src="week3_files/figure-html/box_plot_4-1.png" width="672" /></p>
<p>Displaying points in this way makes much more sense when we only have a few observations and where the box plot masks the fact, perhaps giving the false impression that the sample size is larger than it actually is. Here it makes less sense as we have very many observations.</p>
<p>Let’s try a colour aesthetic to also look at how estrogen receptor expression differs between HER2 positive and negative tumours.</p>
<pre class="r"><code>ggplot(data = metabric, mapping = aes(x = ER_status, y = GATA3, colour = HER2_status)) +
  geom_boxplot()</code></pre>
<p><img src="week3_files/figure-html/box_plot_5-1.png" width="672" /></p>
</div>
<div id="time-series-data" class="section level1">
<h1>Time series data</h1>
<p>The next type of plot we’re going to look at is a <strong>line plot</strong>. Line plots are a common visualization for time series data but we don’t have data of this kind in the METABRIC data set so we’ll turn our attention to a more topical data set.</p>
<div id="covid-19-data-set" class="section level2">
<h2>COVID-19 data set</h2>
<p>Every day the UK Government presents a series of plots showing data relating to the coronavirus (COVID-19) pandemic. Both the plots and the data are published on the Government’s <a href="https://www.gov.uk/government/collections/slides-and-datasets-to-accompany-coronavirus-press-conferences">website</a>.</p>
<p>One of the plots shows the reduction in different types oftransport use over time in the UK in the week leading up to and since the lockdown restrictions were introduced in response to the pandemic.</p>
<p>Reading these data into R from the downloaded Excel spreadsheet requires a bit more care as the spreadsheet contains multiple sheets and each table has header lines and footnotes. The date column doesn’t have a column heading, so in the code below we’ve skipped that row as well and told <code>read_excel()</code> what column names to use. We’ve also given it a heads-up about which type of data are in each column. Look up the help page for <code>read_excel()</code> for more details about the arguments we’ve used below.</p>
<pre class="r"><code>library(readxl)
transport_use &lt;- read_excel(&quot;data/2020-04-26_COVID-19_Press_Conference_Data.xlsx&quot;,
                            sheet = &quot;Transport use&quot;,
                            col_names = c(&quot;date&quot;, &quot;transport_type&quot;, &quot;percentage&quot;),
                            col_types = c(&quot;date&quot;, &quot;text&quot;, &quot;numeric&quot;),
                            skip = 4,
                            n_max = 186)
transport_use$percentage &lt;- transport_use$percentage * 100
transport_use</code></pre>
<pre><code>## # A tibble: 186 x 3
##    date                transport_type     percentage
##    &lt;dttm&gt;              &lt;chr&gt;                   &lt;dbl&gt;
##  1 2020-03-16 00:00:00 All motor vehicles         98
##  2 2020-03-17 00:00:00 All motor vehicles         89
##  3 2020-03-18 00:00:00 All motor vehicles         83
##  4 2020-03-19 00:00:00 All motor vehicles         82
##  5 2020-03-20 00:00:00 All motor vehicles         81
##  6 2020-03-21 00:00:00 All motor vehicles         73
##  7 2020-03-22 00:00:00 All motor vehicles         69
##  8 2020-03-23 00:00:00 All motor vehicles         69
##  9 2020-03-24 00:00:00 All motor vehicles         49
## 10 2020-03-25 00:00:00 All motor vehicles         42
## # … with 176 more rows</code></pre>
<p>In this data set each row represents an observation of the current traffic volume for various modes of transport (motor vehicle, bus, train, etc.) since 16 March as a percentage of the volume measured during the first week of February.</p>
<p>So let’s get a feel for this data set. Firstly, what modes of transport are included and how many observations do we have for each?</p>
<pre class="r"><code>table(transport_use$transport_type)</code></pre>
<pre><code>## 
## All motor vehicles Bus (excl. London)      National rail 
##                 40                 40                 40 
##          TfL - Bus         TfL - Tube 
##                 34                 32</code></pre>
<p>We’ll focus solely on rail use to start with.</p>
<pre class="r"><code>rail_use &lt;- transport_use[transport_use$transport_type == &quot;National rail&quot;, ]
rail_use</code></pre>
<pre><code>## # A tibble: 40 x 3
##    date                transport_type percentage
##    &lt;dttm&gt;              &lt;chr&gt;               &lt;dbl&gt;
##  1 2020-03-16 00:00:00 National rail        77  
##  2 2020-03-17 00:00:00 National rail        69  
##  3 2020-03-18 00:00:00 National rail        59  
##  4 2020-03-19 00:00:00 National rail        52  
##  5 2020-03-20 00:00:00 National rail        44  
##  6 2020-03-21 00:00:00 National rail        38  
##  7 2020-03-22 00:00:00 National rail        35  
##  8 2020-03-23 00:00:00 National rail        29.0
##  9 2020-03-24 00:00:00 National rail        24  
## 10 2020-03-25 00:00:00 National rail        18  
## # … with 30 more rows</code></pre>
</div>
<div id="dates" class="section level2">
<h2>Dates</h2>
<p>There is something unfamiliar about the <code>date</code> column that needs some explanation. Let’s check the type and class of <code>date</code>.</p>
<pre class="r"><code>typeof(rail_use$date)</code></pre>
<pre><code>## [1] &quot;double&quot;</code></pre>
<pre class="r"><code>class(rail_use$date)</code></pre>
<pre><code>## [1] &quot;POSIXct&quot; &quot;POSIXt&quot;</code></pre>
<p>So the <code>date</code> column is stored as a double and is a suitable continuous variable for a line plot but it’s also got some special properties, afforded to it by its <code>POSIXct</code> class, that allows it to be treated as a date.</p>
<p>The <code>POSIXct</code> class is how R represents dates, where POSIX is a standardization scheme in the world of computers, spanning multiple operating systems and programming languages, and the <code>ct</code> part stands for calendar time. We won’t go into the details here but note that the numeric representation of a date, i.e. the <code>double</code> value, is the number of seconds since 1 January 1970. For example, the date in our first row is:</p>
<pre class="r"><code>rail_use$date[1]</code></pre>
<pre><code>## [1] &quot;2020-03-16 UTC&quot;</code></pre>
<pre class="r"><code>as.double(rail_use$date[1])</code></pre>
<pre><code>## [1] 1584316800</code></pre>
</div>
<div id="line-plot" class="section level2">
<h2>Line plot</h2>
<p>We’re going to plot the volume of traffic on each date with each measurement represented as a point on our graph and observations on successive dates connected by a line. Looking at the geoms available in ggplot2 (see the list of functions available for ggplot2 by navigating to the ggplot2 package in the Packages tab in RStudio), it looks like <strong><code>geom_line</code></strong> is the one we need.</p>
<p>The aesthetics section in the help page for <code>geom_line</code> tells us we need an x and y mapping. Our x aesthetic will be <code>date</code> and the y aesthetic will be <code>percentage</code>.</p>
<pre class="r"><code>ggplot(data = rail_use) +
  geom_line(mapping = aes(x = date, y = percentage))</code></pre>
<p><img src="week3_files/figure-html/line_plot_1-1.png" width="672" /></p>
<p>We can return to our original <code>transport_use</code> table containing data for all transport modes and use another aesthetic, e.g. colour, to plot a line for each.</p>
<pre class="r"><code>ggplot(data = transport_use) +
  geom_line(mapping = aes(x = date, y = percentage, colour = transport_type))</code></pre>
<p><img src="week3_files/figure-html/line_plot_2-1.png" width="672" /></p>
<p>Another aesthetic available for <code>geom_line</code> is <code>linetype</code>.</p>
<pre class="r"><code>ggplot(data = transport_use) +
  geom_line(mapping = aes(x = date, y = percentage, linetype = transport_type))</code></pre>
<p><img src="week3_files/figure-html/line_plot_3-1.png" width="672" /></p>
</div>
</div>
<div id="histograms" class="section level1">
<h1>Histograms</h1>
<p>The geom for creating histograms is, rather unsurprisingly, <code>geom_histogram()</code>.</p>
<pre class="r"><code>ggplot(data = metabric) +
  geom_histogram(mapping = aes(x = Age_at_diagnosis))</code></pre>
<pre><code>## `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</code></pre>
<p><img src="week3_files/figure-html/histogram_plot_1-1.png" width="672" /></p>
<p>The warning message hints at picking a more optimal number of bins by specifying the <code>binwidth</code> argument.</p>
<pre class="r"><code>ggplot(data = metabric) +
  geom_histogram(mapping = aes(x = Age_at_diagnosis), binwidth = 5)</code></pre>
<p><img src="week3_files/figure-html/histogram_plot_2-1.png" width="672" /></p>
<p>Or we can set the number of bins.</p>
<pre class="r"><code>ggplot(data = metabric) +
  geom_histogram(mapping = aes(x = Age_at_diagnosis), bins = 20)</code></pre>
<p><img src="week3_files/figure-html/histogram_plot_3-1.png" width="672" /></p>
<p>These histograms are not very pleasing, aesthetically speaking - how about some better aesthetics?</p>
<pre class="r"><code>ggplot(data = metabric) +
  geom_histogram(mapping = aes(x = Age_at_diagnosis), bins = 20, colour = &quot;darkblue&quot;, fill = &quot;grey&quot;)</code></pre>
<p><img src="week3_files/figure-html/histogram_plot_4-1.png" width="672" /></p>
</div>
<div id="categorical-variables-factors" class="section level1">
<h1>Categorical variables – factors</h1>
<p>Several of the variables in the METABRIC data set are categorical, i.e. they can take one of a limited set of values. Some of these have been read into R as character types (e.g. the 3-gene classifier), other as numerical values, e.g. tumour stage. We also have some binary variables that are essentially categorical variables but with only 2 possible values, e.g. ER status.</p>
<p>In many of the plots given above, ggplot2 has treated character variables as categorical in situtations where a categorical variable is expected. For example, when we displayed points on a scatter plot using different colours for each 3-gene classification, or when we created separate box plots in the same graph for ER positive and negative patients.</p>
<p>But what about when our categorical variable has been read into R as a continuous variable, e.g. <code>Tumour_stage</code>, which is read in as a double type.</p>
<pre class="r"><code>ggplot(data = metabric) +
  geom_point(mapping = aes(x = GATA3, y = ESR1, colour = Tumour_stage))</code></pre>
<p><img src="week3_files/figure-html/scatter_plot_9-1.png" width="672" /></p>
<pre class="r"><code>table(metabric$Tumour_stage)</code></pre>
<pre><code>## 
##   0   1   2   3   4 
##   4 475 800 115   9</code></pre>
<p>Tumour stage has only 5 discrete states but ggplot2 doesn’t know these are supposed to be a restricted set of values and has used a colour scale to show them as if they were continuous. We need to tell R that these are categorical.</p>
<p>R has a special type for categorical variables called a <strong>factor</strong>. Let’s convert our tumour stage variable to a factor using the <code>as.factor()</code> function.</p>
<pre class="r"><code>metabric$Tumour_stage &lt;- as.factor(metabric$Tumour_stage)
metabric[, c(&quot;Patient_ID&quot;, &quot;Tumour_stage&quot;)]</code></pre>
<pre><code>## # A tibble: 1,904 x 2
##    Patient_ID Tumour_stage
##    &lt;chr&gt;      &lt;fct&gt;       
##  1 MB-0000    2           
##  2 MB-0002    1           
##  3 MB-0005    2           
##  4 MB-0006    2           
##  5 MB-0008    2           
##  6 MB-0010    4           
##  7 MB-0014    2           
##  8 MB-0022    2           
##  9 MB-0028    2           
## 10 MB-0035    2           
## # … with 1,894 more rows</code></pre>
<p>R actually stores categorical variables as integers but with some additional metadata about which of the integer values, or ‘levels’, corresponds to each category.</p>
<pre class="r"><code>typeof(metabric$Tumour_stage)</code></pre>
<pre><code>## [1] &quot;integer&quot;</code></pre>
<pre class="r"><code>class(metabric$Tumour_stage)</code></pre>
<pre><code>## [1] &quot;factor&quot;</code></pre>
<pre class="r"><code>levels(metabric$Tumour_stage)</code></pre>
<pre><code>## [1] &quot;0&quot; &quot;1&quot; &quot;2&quot; &quot;3&quot; &quot;4&quot;</code></pre>
<pre class="r"><code>ggplot(data = metabric) +
  geom_point(mapping = aes(x = GATA3, y = ESR1, colour = Tumour_stage))</code></pre>
<p><img src="week3_files/figure-html/scatter_plot_10-1.png" width="672" /></p>
<p>In this case the order of the levels makes sense but for other variables you may wish for more control over the ordering. Take the integrative cluster variable for example. We created a bar plot of the numbers of patients in the METABRIC cohort within each integrative cluster. Did you notice the ordering of the clusters? 10 came just after 1 and before 2. That looked a bit odd as we’d have naturally expected it to come last of all. R, on the other hand, is treating this vector as a character vector (mainly because of the ‘ER-’ and ‘ER+’ subtypes of cluster 4, and sorts the values into alphanumerical order.</p>
<pre class="r"><code>metabric$Integrative_cluster &lt;- as.factor(metabric$Integrative_cluster)
levels(metabric$Integrative_cluster)</code></pre>
<pre><code>##  [1] &quot;1&quot;    &quot;10&quot;   &quot;2&quot;    &quot;3&quot;    &quot;4ER-&quot; &quot;4ER+&quot; &quot;5&quot;    &quot;6&quot;    &quot;7&quot;    &quot;8&quot;   
## [11] &quot;9&quot;</code></pre>
<p>A look at the help page for <code>as.factor</code> shows that we can create a factor using the <code>factor()</code> function and specify the levels using the <code>levels</code> argument.</p>
<pre class="r"><code>metabric$Integrative_cluster &lt;- factor(metabric$Integrative_cluster, levels = c(&quot;1&quot;, &quot;2&quot;, &quot;3&quot;, &quot;4ER-&quot;, &quot;4ER+&quot;, &quot;5&quot;, &quot;6&quot;, &quot;7&quot;, &quot;8&quot;, &quot;9&quot;, &quot;10&quot;))
levels(metabric$Integrative_cluster)</code></pre>
<pre><code>##  [1] &quot;1&quot;    &quot;2&quot;    &quot;3&quot;    &quot;4ER-&quot; &quot;4ER+&quot; &quot;5&quot;    &quot;6&quot;    &quot;7&quot;    &quot;8&quot;    &quot;9&quot;   
## [11] &quot;10&quot;</code></pre>
<pre class="r"><code>ggplot(data = metabric) +
  geom_bar(mapping = aes(x = Integrative_cluster, fill = Integrative_cluster))</code></pre>
<p><img src="week3_files/figure-html/bar_plot_9-1.png" width="672" /></p>
</div>
<div id="the-ggplot-object-a-peek-under-the-hood-advanced" class="section level1">
<h1>The ggplot object – a peek under the hood (ADVANCED)</h1>
<p>Let’s build a ggplot2 plot up in stages to understand what’s really going on.</p>
<pre class="r"><code>plot &lt;- ggplot(data = metabric)</code></pre>
<p>What is the plot object we’ve just created?</p>
<pre class="r"><code>typeof(plot)</code></pre>
<pre><code>## [1] &quot;list&quot;</code></pre>
<pre class="r"><code>length(plot)</code></pre>
<pre><code>## [1] 9</code></pre>
<p>The plot object is a list containing 9 elements. It’s actually a special type of list.</p>
<pre class="r"><code>class(plot)</code></pre>
<pre><code>## [1] &quot;gg&quot;     &quot;ggplot&quot;</code></pre>
<p>What are the 9 things in the list?</p>
<pre class="r"><code>names(plot)</code></pre>
<pre><code>## [1] &quot;data&quot;        &quot;layers&quot;      &quot;scales&quot;      &quot;mapping&quot;     &quot;theme&quot;      
## [6] &quot;coordinates&quot; &quot;facet&quot;       &quot;plot_env&quot;    &quot;labels&quot;</code></pre>
<p>So far we’ve constructed ggplots from data, layers (geoms) and aesthetic mappings but do you notice the labels, scales, facet and theme elements? We’ll be looking at how to customize plots later, including changing the labels, altering the scales of axes, and applying a different visual theme.</p>
<p>The <code>data</code> element is just the metabric tibble that we provided to the <code>ggplot()</code> function.</p>
<pre class="r"><code>plot$data</code></pre>
<pre><code>## # A tibble: 1,904 x 30
##    Patient_ID Cohort Age_at_diagnosis Survival_time Survival_status
##    &lt;chr&gt;       &lt;dbl&gt;            &lt;dbl&gt;         &lt;dbl&gt; &lt;chr&gt;          
##  1 MB-0000         1             75.6         140.  LIVING         
##  2 MB-0002         1             43.2          84.6 LIVING         
##  3 MB-0005         1             48.9         164.  DECEASED       
##  4 MB-0006         1             47.7         165.  LIVING         
##  5 MB-0008         1             77.0          41.4 DECEASED       
##  6 MB-0010         1             78.8           7.8 DECEASED       
##  7 MB-0014         1             56.4         164.  LIVING         
##  8 MB-0022         1             89.1          99.5 DECEASED       
##  9 MB-0028         1             86.4          36.6 DECEASED       
## 10 MB-0035         1             84.2          36.3 DECEASED       
## # … with 1,894 more rows, and 25 more variables: Vital_status &lt;chr&gt;,
## #   Chemotherapy &lt;chr&gt;, Radiotherapy &lt;chr&gt;, Tumour_size &lt;dbl&gt;,
## #   Tumour_stage &lt;fct&gt;, Neoplasm_histologic_grade &lt;dbl&gt;,
## #   Cancer_type &lt;chr&gt;, ER_status &lt;chr&gt;, PR_status &lt;chr&gt;,
## #   HER2_status &lt;chr&gt;, HER2_status_measured_by_SNP6 &lt;chr&gt;, PAM50 &lt;chr&gt;,
## #   `3-gene_classifier` &lt;chr&gt;, Nottingham_prognostic_index &lt;dbl&gt;,
## #   Cellularity &lt;chr&gt;, Integrative_cluster &lt;fct&gt;, Mutation_count &lt;dbl&gt;,
## #   ESR1 &lt;dbl&gt;, ERBB2 &lt;dbl&gt;, PGR &lt;dbl&gt;, TP53 &lt;dbl&gt;, PIK3CA &lt;dbl&gt;,
## #   GATA3 &lt;dbl&gt;, FOXA1 &lt;dbl&gt;, MLPH &lt;dbl&gt;</code></pre>
<p>This <code>plot</code> list object is really only a specification for how ggplot2 should create the plot. ggplot2 only renders the plot when we ask it to by printing it out to the screen by typing <code>print(plot)</code> or simply just <code>plot</code>.</p>
<p>We haven’t added any layers yet so what does our plot look like at this point?</p>
<pre class="r"><code>plot</code></pre>
<p><img src="week3_files/figure-html/building_up_plot_1-1.png" width="672" /></p>
<p>ggplot2 doesn’t know what to plot yet as we haven’t added a layer (geom).</p>
<pre class="r"><code>plot$mapping</code></pre>
<pre><code>## Aesthetic mapping: 
## &lt;empty&gt;</code></pre>
<pre class="r"><code>plot$layers</code></pre>
<pre><code>## list()</code></pre>
<p>Lets add an aesthetic mapping.</p>
<pre class="r"><code>plot &lt;- ggplot(data = metabric, mapping = aes(x = Nottingham_prognostic_index, y = ESR1, colour = ER_status))
plot</code></pre>
<p><img src="week3_files/figure-html/building_up_plot_2-1.png" width="672" /></p>
<p>ggplot2 has automatically added scales for x and y based on the ranges of values for the Nottingham prognostic index and ESR1 expression. Still nothing has been plotted as we haven’t yet specified the type of plot (geom) to add as a layer.</p>
<pre class="r"><code>plot$mapping</code></pre>
<pre><code>## Aesthetic mapping: 
## * `x`      -&gt; `Nottingham_prognostic_index`
## * `y`      -&gt; `ESR1`
## * `colour` -&gt; `ER_status`</code></pre>
<pre class="r"><code>plot$layers</code></pre>
<pre><code>## list()</code></pre>
<p>Finally, let’s add a <code>geom_point</code> layer to create a scatter plot.</p>
<pre class="r"><code>plot &lt;- plot + geom_point()
plot</code></pre>
<p><img src="week3_files/figure-html/building_up_plot_3-1.png" width="672" /></p>
<pre class="r"><code>plot$layers</code></pre>
<pre><code>## [[1]]
## geom_point: na.rm = FALSE
## stat_identity: na.rm = FALSE
## position_identity</code></pre>
<p>We touched on statistical transformations earlier. The <code>stat</code> associated with a <code>geom_point</code> is <code>stat_identity</code> which leaves values unchanged. In the case of a scatter plot, we already have the x and y values – they don’t need to be transformed, just plotted on the x and y axes.</p>
<hr />
</div>
<div id="summary" class="section level1">
<h1>Summary</h1>
<p>In this session we have covered the following concepts:</p>
<ul>
<li>How to use ggplot2 to create different types of plot using a common grammar</li>
<li>How to map variables to aesthetic features of a plot</li>
<li>How to construct a plot with multiple layers (geoms)</li>
<li>Various types of plot including scatter plots, box plots, bar plots and histograms</li>
<li>Working with dates</li>
<li>Plotting time series data as line graphs</li>
<li>How to work with time series data</li>
<li>Categorical variables and how R handles these with its factor type</li>
</ul>
<hr />
</div>
<div id="assignment" class="section level1">
<h1>Assignment</h1>
<p>Assignment: <a href="assignments/assignment3.Rmd">assignment3.Rmd</a></p>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
